numpy/_core/src/multiarray/stringdtype/static_string.c

```diff
@@ -0,0 +1,742 @@
+}
+
+static int
+is_a_vstring(const npy_packed_static_string *s)
+{
+    return !is_not_a_vstring(s);
+}
+
+/*NUMPY_API
+ * Acquire the mutex locking the allocator attached to *descr*.
+ *
+ * NpyString_release_allocator must be called on the allocator returned
+ * by this function exactly once.
+ */
+npy_string_allocator *
+NpyString_acquire_allocator(PyArray_StringDTypeObject *descr)
+{
+    if (!PyThread_acquire_lock(descr->allocator->allocator_lock, NOWAIT_LOCK)) {
+        PyThread_acquire_lock(descr->allocator->allocator_lock, WAIT_LOCK);
+    }
+    return descr->allocator;
+}
+
+/*NUMPY_API
+ * Simultaneously acquire the mutexes locking the allocators attached to
+ * *N* descriptors. Note that the descrs array can have more elements
+ * than n_allocators, and any non-stringdtype descriptors are skipped,
+ * allowing passing the `context->descriptors` array inside a ufunc or cast
+ * loop.
+ *
+ * There is no error checking for setting n_allocators too large and buffer
+ * overflows might happen if there are not n_allocators StringDType descriptors
+ * in the descrs array.
+ *
+ * If pointers to the same descriptor are passed multiple times, only acquires
+ * the allocator mutex once but sets identical allocator pointers appropriately.
+ *
+ * The allocator mutexes must be released after this function returns, see
+ * NpyString_release_allocators.
+ */
+void
+NpyString_acquire_allocators(size_t n_allocators,
+                             PyArray_Descr *descrs[],
+                             npy_string_allocator *allocators[])
+{
+    size_t allocator_count = 0;
+    for (size_t i=0; allocator_count<n_allocators; i++) {
+        if (NPY_DTYPE(descrs[i]) != &PyArray_StringDType) {
+            continue;
+        }
+        int descriptors_match = 0;
+        for (size_t j=0; j<i; j++) {
+            if (descrs[i] == descrs[j]) {
+                allocators[allocator_count] = allocators[j];
+                descriptors_match = 1;
+                break;
+            }
+        }
+        if (!descriptors_match) {
+            allocators[allocator_count] = NpyString_acquire_allocator(
+                    (PyArray_StringDTypeObject *)descrs[i]);
+        }
+        allocator_count += 1;
+    }
+    assert(allocator_count == n_allocators);
+}
+
+/*NUMPY_API
+ * Release the mutex locking an allocator. This must be called exactly once
+ * after acquiring the allocator mutex and all operations requring the
+ * allocator are done.
+ * 
+ * If you need to release multiple allocators, see
+ * NpyString_release_allocators, which can correctly handle releasing the
+ * allocator once when several references are passed in.
+ */
+void
+NpyString_release_allocator(npy_string_allocator *allocator)
+{
+    PyThread_release_lock(allocator->allocator_lock);
+}
+
+/*NUMPY_API
+ * Release the mutexes locking N allocators.  If pointers to the same
+ * allocator are passed multiple times, only releases the allocator mutex once.
+ */
+void
```

Comment:
Just a thought, it might be a fun sprint activity for someone to add a test that uses e.g. treesitter to check that every function that should have an annotation does have one.